### PR TITLE
Replace NihHash usage in trace.c with own set implementation

### PR DESCRIPTION
**Pull request by komugi1211s**
This is part of ongoing effort to decouple libnih from the ureadahead.

During tracing process, ureadahead utilizes the hashmap from libnih to prevent duplicate files from accidentally getting processed twice.
This PR replaces said hashmap with our own implementation of hash set.

### CODE REVIEW

**Comment by rostedt** [comment]
Can you rebase. I would fix the conflicts but the other branch adds another use of NihHash, where that needs to be replaced too. I pulled your other branch in that conflicts.

**Comment by komugi1211s** [comment]
@rostedt Hi, Sorry for the late ping.
I have rebased the branch.

File: src/trace.c [hunk 1]
```diff
@@ -2121,3 +2140,79 @@ static struct device_data *find_device (struct device_data **device_hash,
...
+	data = calloc (1, sizeof (struct path_data));
+	assert (data != NULL);
+
+	data->hash = hash;
+	data->length = length;
+
+	strncpy (data->path, pathname, PACK_PATH_MAX);
+	data->path[PACK_PATH_MAX] = '\0';
+
+	if (path_table[key])
```
Comment (rostedt) [inline src/trace.c]: You are not saving anything with this compare. Control flow can actually be more expensive than an assignment to hot cache. Just assign `data->next = path_table[key];`. It makes the code easier to read as well.

File: src/trace.c [hunk 2]
```diff
@@ -2121,3 +2140,79 @@ static struct device_data *find_device (struct device_data **device_hash,
...
+	/* dev/inode pair not found. add a new one,
+	 * and insert it to the start of linked list.
+	 */
+	data = calloc (1, sizeof (struct dev_inode_data));
+	assert (data != NULL);
+
+	data->dev_id = dev_id;
+	data->inode  = inode;
+
+	if (dev_inode_table[key])
```
Comment (rostedt) [inline src/trace.c]: Same here.

**Pull request merged**